src/bin/common/solving_loop.ml

```diff
@@ -834,79 +937,298 @@ let main () =
+          in
+          let hyps, t =
+            match contents with
+            | `Goal t ->
+              D_cnf.pp_query t
+            | `Check hyps ->
+              D_cnf.pp_query ~hyps (DStd.Expr.Term.(of_cst Const._false))
+          in
+          let () =
+            List.iter (
+              fun t ->
+                let name = Ty.fresh_hypothesis_name goal_sort in
+                assume_axiom st name t loc attrs
+            ) hyps
+          in
+          let e = D_cnf.make_form "" t st_loc ~decl_kind:Expr.Dgoal in
+          (* Performing the query. *)
+          Api.FE.query ~loc:st_loc (name, e, goal_sort) Api.env;
+          (* Treatment of the result. *)
+          let partial_model = Api.env.sat_env in
+          (* If the status of the SAT environment is inconsistent,
+             we have to drop the partial model in order to prevent
+             printing wrong model. *)
+          match Api.env.res with
+          | `Sat ->
+            begin
+              let mdl = Model ((module Api.SAT), partial_model) in
+              let () =
+                if Options.(get_interpretation () && get_dump_models ()) then
+                  Api.FE.print_model
+                    (Options.Output.get_fmt_models ())
+                    partial_model
+              in
+              Sat mdl
+            end
+          | `Unknown ->
+            begin
+              let mdl = Model ((module Api.SAT), partial_model) in
+              if Options.(get_interpretation () && get_dump_models ()) then
+                begin
+                  let ur = Api.SAT.get_unknown_reason partial_model in
+                  Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+                    "@[<v 0>Returned unknown reason = %a@]"
+                    Sat_solver_sig.pp_ae_unknown_reason_opt ur;
+                  Api.FE.print_model
+                    (Options.Output.get_fmt_models ())
+                    partial_model
+                end;
+              Unknown (Some mdl)
+            end
+          | `Unsat -> Unsat Api.env.expl
+        end
+    in
+    (* Prints the result. *)
+    print_solve_res st_loc name solve_res;
+    (* Updates the dolmen state. *)
+    set_partial_model_and_mode solve_res st
+  in
+
   let handle_stmt :
-    Frontend.used_context -> State.t ->
-    'a D_loop.Typer_Pipe.stmt -> State.t =
+    Frontend.used_context ->
+    State.t ->
+    [< Typer_Pipe.typechecked | `Check of 'a ] D_loop.Typer_Pipe.stmt ->
+    State.t =
     let goal_cnt = ref 0 in
-    fun all_context st td ->
+    fun
+      _all_context
+      st
+      td ->
       let file_loc = (State.get State.logic_file st).loc in
-      let solver_ctx = State.get solver_ctx_key st in
       match td with
+      (* Set logic *)
       | { contents = `Set_logic _; _} ->
         cmd_on_modes st [Start] "set-logic";
         DO.Mode.set Util.Assert st
+
+      (* Push and Pop commands *)
+      | { contents = `Pop n; loc; _ } ->
```

Comment:
This still needs to update the `incremental_depth` for the optimization compatibility check (see #993 ).